src/shapirowilk.jl

```diff
@@ -0,0 +1,248 @@
+export ShapiroWilkTest
+
+#=
+From:
+PATRICK ROYSTON
+Approximating the Shapiro-Wilk W-test for non-normality
+*Statistics and Computing* (1992) **2**, 117-119
+DOI: [10.1007/BF01891203](https://doi.org/10.1007/BF01891203)
+=#
+
+# TODO: Rerun simulation and polynomial fitting
+
+# Coefficients from Royston (1992)
+for (s, c) in [(:C1, [0.0, 0.221157, -0.147981, -2.07119, 4.434685, -2.706056]),
+               (:C2, [0.0, 0.042981, -0.293762, -1.752461, 5.682633, -3.582633]),
+               (:C3, [0.5440, -0.39978, 0.025054, -0.0006714]),
+               (:C4, [1.3822, -0.77857, 0.062767, -0.0020322]),
+               (:C5, [-1.5861, -0.31082, -0.083751, 0.0038915]),
+               (:C6, [-0.4803, -0.082676, 0.0030302]),
+               (:C7, [0.164, 0.533]),
+               (:C8, [0.1736, 0.315]),
+               (:C9, [0.256, -0.00635]),
+               (:G, [-2.273, 0.459])]
+    @eval $(Symbol(:__RS92_, s))(x) = Base.Math.@horner(x, $(c...))
+end
+
+#=
+The following hardcoded constants has been replaced by more precise values:
+
+SQRTH = sqrt(2.0)/2.0 # 0.70711E0
+TH = 3/8 # 0.375E0
+SMALL = eps(1.0) # 1E-19
+PI6 = π/6 # 0.1909859E1
+STQR = asin(sqrt(0.75)) # 0.1047198E1
+=#
+
+struct SWCoeffs <: AbstractVector{Float64}
+    N::Int
+    A::Vector{Float64}
+end
+
+Base.size(SWc::SWCoeffs) = (SWc.N,)
+Base.IndexStyle(::Type{SWCoeffs}) = IndexLinear()
+
+Base.@propagate_inbounds function Base.getindex(SWc::SWCoeffs, i::Integer)
+    @boundscheck checkbounds(SWc, i)
+    @inbounds if checkbounds(Bool, SWc.A, i)
+        return SWc.A[i]
+    elseif isodd(length(SWc)) && i == lastindex(SWc.A) + 1
+        return zero(eltype(SWc))
+    else
+        return -SWc.A[SWc.N+1-i]
+    end
+end
+
+function SWCoeffs(N::Int)
+    if N < 3
+        throw(ArgumentError("N must be greater than or equal to 3: $N"))
+    elseif N == 3 # exact
+        return SWCoeffs(N, [sqrt(2.0) / 2.0])
+    else
+        # Weisberg&Bingham 1975 statistic; store only positive half of m:
+        # it is (anti-)symmetric; hence '2' factor below
+        m = [-quantile(Normal(), (i - 3 / 8) / (N + 1 / 4)) for i in 1:div(N, 2)]
+        mᵀm = 2sum(abs2, m)
+        x = 1 / sqrt(N)
+        a₁ = m[1] / sqrt(mᵀm) + __RS92_C1(x) # aₙ = cₙ + (...)
+        if N ≤ 5
+            ϕ = (mᵀm - 2m[1]^2) / (1 - 2a₁^2)
+            m .= m ./ sqrt(ϕ) # A, but reusing m to save allocs
+            m[1] = a₁
+        else
+            a₂ = m[2] / sqrt(mᵀm) + __RS92_C2(x) # aₙ₋₁ = cₙ₋₁ + (...)
+            ϕ = (mᵀm - 2m[1]^2 - 2m[2]^2) / (1 - 2a₁^2 - 2a₂^2)
+            m .= m ./ sqrt(ϕ) # A, but reusing m to save allocs
+            m[1], m[2] = a₁, a₂
+        end
+
+        return SWCoeffs(N, m)
+    end
+end
+
+function swstat(X::AbstractArray{<:Real}, A::SWCoeffs)
+    if last(X) - first(X) < length(X) * eps()
+        throw(ArgumentError("sample seems to be constant!"))
```

Comment:
Saying "seems" is a bit weird here: can we say "is" constant? Not sure whether the tolerance has to be mentioned.

(Also no need for the exclamation mark.)

Reply:
done